src/apps/activities/api/activities.py

```diff
@@ -358,3 +360,97 @@ async def __filter_activities(
+    applet_service = AppletService(session, user.id)
+    await applet_service.exist_by_id(applet_id)
+    await CheckAccessService(session, user.id).check_applet_detail_access(applet_id)
+
+    await SubjectsService(session, user.id).exist_by_id(subject_id)
+
+    # Fetch assigned activity or flow IDs for the subject
+    assigned_activities = await ActivityAssignmentService(session).get_assigned_activity_or_flow_ids_for_subject(
+        subject_id
+    )
+
+    # Fetch activities submissions by the subject
+    submitted_activities = await AnswerService(session, user.id, answer_session).get_submissions_by_subject(subject_id)
+
+    # Fetch auto-assigned activity and flow IDs by applet ID
+    auto_activity_ids = await ActivityService(session, user.id).get_activity_and_flow_ids_by_applet_id_auto(applet_id)
+
+    # Combine all assigned IDs and submitted activity IDs
+    all_activity_ids = (
+        set(assigned_activities.activities.keys())
+        | set(submitted_activities.activities.keys())
+        | set(auto_activity_ids)
+    )
+
+    # Initialize ActivitiesCounters with zero counts
+    activities_counters = ActivitiesCounters(
+        subject_id=subject_id, respondent_activities_count=0, target_activities_count=0, activities_or_flows=[]
+    )
+
+    # Iterate over all activity or flow IDs
+    for activity_or_flow_id in all_activity_ids:
+        is_auto = activity_or_flow_id in auto_activity_ids
+
+        # Get submission and assignment data if available
+        submission_data = submitted_activities.activities.get(activity_or_flow_id)
+        assignments_data = assigned_activities.activities.get(activity_or_flow_id)
+
+        # Initialize sets for respondents and subjects
+        respondents = set()
+        subjects = set()
+
+        # Initialize submission counts
+        respondent_submissions_count = 0
+        subject_submissions_count = 0
+
+        # Update from submission data
+        if submission_data:
+            respondents.update(submission_data.respondents)
+            subjects.update(submission_data.subjects)
+            respondent_submissions_count = submission_data.respondent_submissions_count
+            subject_submissions_count = submission_data.subject_submissions_count
+
+        # Update from assignment data
+        if assignments_data:
+            respondents.update(assignments_data.respondents)
+            subjects.update(assignments_data.subjects)
+
+        # Include the subject for auto-assigned activities
+        if is_auto:
+            respondents.add(subject_id)
+            subjects.add(subject_id)
+
+        # Calculate counts
+        respondents_count = len(respondents)
+        subjects_count = len(subjects)
+
+        # Update activities counters counts
+        if subjects_count > 0:
+            activities_counters.respondent_activities_count += 1
+        if respondents_count > 0:
+            activities_counters.target_activities_count += 1
```

Comment:
So, the next issue I'm running into with the counts is with the `activities_counters`. In your PR, activities and flows that have been deleted (and are now only present in the `activity_histories`/`flow_histories` tables) are incrementing the total activities/flows here. But in the ticket, the last AC bullet says:

> **Out of Scope:**
>
> Future updates should include provisions for handling deleted activities/flows as per [M2-7908](https://mindlogger.atlassian.net/browse/M2-7908).

Since we're not listing any deleted activities or flows in the PDP pages yet, we also need the total counts to reflect that fact. However, _eventually_ we will want to include the deleted activities/flows in the counts, once M2-7908 is completed.

Instead of returning aggregate `respondent_activities_count` and `target_activities_count` values to the response that includes both existing and deleted activities/flows, can we instead return more granular values, and the FE can decide which counts to display (based on eventual support for deleted activities/flows):

```json
{
    "result": {
        "subjectId": …,
        "respondentActivitiesCountExisting": 0,
        "respondentActivitiesCountDeleted": 0,
        "targetActivitiesCountExisting": 0,
        "targetActivitiesCountDeleted": 0,
        "activitiesOrFlows": [ … ]
    }
}
```

Reply:
FYI for this change, we only need the top-level counts to be able to distinguish how many activities/flows are deleted. The `activitiesOrFlows` array data need not change.